cpp/daal/src/externals/service_rng_ref.h

```diff
@@ -0,0 +1,363 @@
+    virtual int uniformRNG(const size_t n, int * r, const int a, const int b, const int method)               = 0;
+    virtual int uniformRNG(const size_t n, float * r, const float a, const float b, const int method)         = 0;
+    virtual int uniformRNG(const size_t n, double * r, const double a, const double b, const int method)      = 0;
+    virtual int gaussianRNG(const size_t n, float * r, const float a, const float sigma, const int method)    = 0;
+    virtual int gaussianRNG(const size_t n, double * r, const double a, const double sigma, const int method) = 0;
+    virtual int bernoulliRNG(const size_t n, int * r, const double p, const int method)                       = 0;
+    virtual int getSize()                                                                                     = 0;
+    virtual void clone(void * dest) const                                                                     = 0;
+    virtual StateIface * clone() const                                                                        = 0;
+    virtual void assign(const void * src)                                                                     = 0;
+    virtual void discard(size_t nSkip)                                                                        = 0;
+};
+
+template <typename Gen = std::mt19937>
+class State : public StateIface
+{
+    using ThisType = State<Gen>;
+
+public:
+    int uniformRNG(const size_t n, size_t * r, const size_t a, const size_t b, const int method) override { return iUniform(n, r, a, b, method); }
+
+    int uniformRNG(const size_t n, int * r, const int a, const int b, const int method) override { return iUniform(n, r, a, b, method); }
+
+    int uniformRNG(const size_t n, float * r, const float a, const float b, const int method) override { return fUniform(n, r, a, b, method); }
+
+    int uniformRNG(const size_t n, double * r, const double a, const double b, const int method) override { return fUniform(n, r, a, b, method); }
+
+    int gaussianRNG(const size_t n, float * r, const float a, const float sigma, const int method) override
+    {
+        return fGaussian(n, r, a, sigma, method);
+    }
+
+    int gaussianRNG(const size_t n, double * r, const double a, const double sigma, const int method) override
+    {
+        return fGaussian(n, r, a, sigma, method);
+    }
+
+    int bernoulliRNG(const size_t n, int * r, const double p, const int method) override { return iBernoulli(n, r, p, method); }
+
+    State(const int brngId, const size_t seedSize, const unsigned int * seed)
+        : _gen(0), _brngId(brngId), _seedSize(seedSize), _seed(nullptr), _nSkip(0)
+    {
+        _seed = (unsigned int *)daal::services::daal_malloc(sizeof(unsigned int) * _seedSize);
+        if (_seedSize > 0 && seed != nullptr)
+        {
+            for (size_t i = 0; i < _seedSize; i++)
+            {
+                _seed[i] = seed[i];
+            }
+            _gen.seed(_seed[0]);
+        }
+    }
```

Comment:
Please use the way that is described in C++ standard to serialize/deserialize C++ generators.
It says that `operator<<` and `operator>>` are defined for this purpose.

The example below shows how to save C++ RNG's state into the buffer, and then how to load it from the buffer to create an identical RNG:
```cpp
#include <iostream>
#include <sstream>
#include <cstring>
#include <random>

int main() {
    std::mt19937 gen1(777);

    std::stringstream ss_serialized;
    ss_serialized << gen1;

    const int state_lenght = ss_serialized.str().length();
    char * buffer = new char[state_lenght];

    memcpy(buffer, ss_serialized.str().c_str(), state_lenght);

    std::cout << buffer << std::endl;

    std::stringstream ss_deserialized;

    ss_deserialized << buffer;
    

    std::mt19937 gen2(0);
    ss_deserialized >> gen2;

    std::uniform_real_distribution<> dis(1.0, 2.0);
    std::cout << "Uniform distribution + gen1:" << std::endl;
    for(int i = 0; i < 10; ++i)
    {
        std::cout << dis(gen1) << ' ';
    }
    std::cout << std::endl;
    std::cout << "Uniform distribution + gen2:" << std::endl;
    for(int i = 0; i < 10; ++i)
    {
        std::cout << dis(gen2) << ' ';
    }
    std::cout << std::endl;
    delete[] buffer;
}
```